ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -622,6 +622,24 @@ static const upb_pbdecodermethod *msgdef_decodermethod(Descriptor* desc) {
   return desc->fill_method;
 }
 
+// Closure of information required within an encode/decode-error callback: the
+// environment to free, and the error message template to use when raising a
+// Ruby exception.
+typedef struct {
+  upb_env* env;
+  const char* msg;
+} env_errdata;
+
+// Callback invoked by upb if any error occurs during parsing or serialization.
+static bool env_error_func(void* ud, const upb_status* status) {
+  env_errdata* errdata = ud;
+  // Free the env -- rb_raise will longjmp up the stack past the encode/decode
+  // function so it would not otherwise have been freed.
+  upb_env_uninit(errdata->env);
```

Comment:
I guess here you're exercising another corner: is it safe to call upb_env_uninit() from inside an error callback? As long as you are longjmp-ing across any encoders/decoders on the stack, it should probably be ok. But this pattern would *require* that you immediately longjmp().

So your options for longjmp are either:

1. longjmp() to a frame that has the upb_env still in scope, then call upb_env_uninit().
2. call upb_env_uninit() then *immediately* longjmp() past any encoders/decoders on the CPU stack, and do not touch the env again. A regular return from the function in this case would cause a crash!

Both of these seem like supportable patterns.